c++/src/kj/compat/tls.c++

```diff
@@ -407,29 +420,30 @@ private:
 // =======================================================================================
 // Implementations of ConnectionReceiver, NetworkAddress, and Network as wrappers adding TLS.
 
-class TlsConnectionReceiver final: public kj::ConnectionReceiver {
+class TlsConnectionReceiver final: public ConnectionReceiver, public TaskSet::ErrorHandler {
 public:
-  TlsConnectionReceiver(TlsContext& tls, kj::Own<kj::ConnectionReceiver> inner)
-      : tls(tls), inner(kj::mv(inner)) {}
+  TlsConnectionReceiver(TlsContext& tls, Own<ConnectionReceiver> inner)
+      : tls(tls), inner(kj::mv(inner)), tasks(*this) {
+    // Start up the connection loop.
+    tasks.add(doOneAccept());
+  }
+
+  void taskFailed(Exception&& e) override {
+    KJ_LOG(ERROR, "error accepting tls connection", kj::fwd<Exception>(e));
+  };
 
   Promise<Own<AsyncIoStream>> accept() override {
-    return inner->accept().then([this](kj::Own<AsyncIoStream> stream) {
-      return kj::evalNow([&] { return tls.wrapServer(kj::mv(stream)); })
-          .catch_([this](kj::Exception&& e) {
-        KJ_LOG(ERROR, "error accepting tls connection", e);
-        return accept();
-      });
+    return acceptAuthenticated().then([](AuthenticatedStream&& stream) {
+      return kj::mv(stream.stream);
     });
   }
 
-  Promise<kj::AuthenticatedStream> acceptAuthenticated() override {
-    return inner->acceptAuthenticated().then([this](kj::AuthenticatedStream stream) {
-      return kj::evalNow([&] { return tls.wrapServer(kj::mv(stream)); })
-          .catch_([this](kj::Exception&& e) {
-        KJ_LOG(ERROR, "error accepting tls connection", e);
-        return acceptAuthenticated();
-      });
-    });
+  Promise<AuthenticatedStream> acceptAuthenticated() override {
+    return queue.pop();
```

Comment:
If a fatal error occurred, then `accept()` should throw it, since that's what all existing uses of this class expect. You could do:

```
return queue.pop().exclusiveJoin(onFatalError());
```

Reply:
Hmmmm, I don't think that works because the Promises have separate types. I can think of a few different solutions but most of them aren't so great. :/ Dropping a comment in place of a better idea for now.

Reply:
Since `onFatalError()` never resolves, only throws, you can do:

```
return queue.pop().exclusiveJoin(
    onFatalError().then([]() -> AuthenticatedStream { KJ_UNREACHABLE; }));
```